### DISCUSSION TITLE
Modelling cylindric lenses

### DISCUSSION THREAD

**Discussion by JanHett** [mjhoptics/ray-optics#95] [Q&A] [answered]
Hello!

Is it possible to model cylindric surfaces (x- or y-aligned as well as rotated around z by a given angle) with rayoptics? Specifically, I am trying to model a system inspired by the Wallin patent US2890622A (two counter-rotating cylindric lenses used to correct the astigmatism found when focusing a spherical lens behind an anamorphoser). Digging through the code I couldn't find anything that pointed me into the right direction so I figured I'd ask before attempting to knit my own.

Thanks a lot in advance and have an excellent day!

**Comment by mjhoptics** [comment, marked as answer]
Hello Jan,

x and y toroid surfaces are available in the [profiles](https://ray-optics.readthedocs.io/en/stable/api/rayoptics.elem.profiles.html#module-rayoptics.elem.profiles) module. You can set the sweep radius to infinity (sweep curvature to 0) and then you have a cylinder. The [YToroid](https://ray-optics.readthedocs.io/en/stable/api/rayoptics.elem.profiles.html#rayoptics.elem.profiles.YToroid) can be used as a Y Cylinder by setting `cR` to 0. There is a XToroid that can be used similarly. 

To do intermediate angles, I would recommend using gamma rotations on the parts with the cylinder surfaces. The [DecenterData](https://ray-optics.readthedocs.io/en/stable/api/rayoptics.elem.surface.html#rayoptics.elem.surface.DecenterData) is an attribute of the [Surface](https://ray-optics.readthedocs.io/en/stable/api/rayoptics.elem.surface.html#rayoptics.elem.surface.Surface) and is specified by the surface offsets and euler angle rotations. A gamma Euler angle rotation can be used to "clock" an element. 

Hope this helps; feel free to follow up. Mike

**Reply by JanHett** [reply]
Hello Mike,

Thanks a lot! That was exactly what I was looking for.

Am I seeing this right that the rotation isn't represented when plotting the model? Not a huge issue if that's indeed not the case, since it would be just to help me with visualisation.

**Reply by mjhoptics** [reply]
Yes, you are right that the model plot doesn't include the gamma rotation. Someday I'll have to implement a 3D view...

**Comment by khoroshyy** [comment]
Hi Michael. @mjhoptics 
I have realized here is a better place for discussion. 

I have trouble realize what is the difference between XToroid and YToroid. Initially, I thought it is the orientation of the cylinder. But the raytracing of the XToroid and YToroid lens were the same.  I would expect one to work as a lens and another as a glass plate. Or am I missing something? 
Thanks.

Petro.

**Reply by mjhoptics** [reply]
Petro,
There was a bug and I fixed it. I will create a new release in the next day or so. Thank you for reporting it. I am attaching a pdf of the notebook I used to test the fix.
Thanks again.
Mike
[Issue95 XToroid raytrace.pdf](https://github.com/mjhoptics/ray-optics/files/9702297/Issue95.XToroid.raytrace.pdf)

**Reply by mikejaris** [reply]
Hi Mike @mjhoptics ,

First off, thank you so much for creating ray-optics! I am amazed by the amount of effort you've put into this project, it works wonderfully  and I've learned a lot playing around with it.

I am resurrecting this thread because I've encountered some unexpected behavior using XToroid. I've attached a .pdf of a jupyter notebook where I compare the spot diagrams of optical systems using YToroid and XToroid objects. YToroid produces the result I would expect (light is focused in one dimension), but XToroid produces a very different result. I would expect the spot diagram to look like a 90 degree rotation of the YToroid result, instead, it yields a skewed pattern that appears to be rotated only ~45 degrees and does not focus as tightly as the YToroid example. 

Do these results make sense to you (i.e., I'm doing something wrong) or do you think there is an error with the XToroid class?

Cheers,
Mike
[CylindricalLensTesting.pdf](https://github.com/mjhoptics/ray-optics/files/15069704/CylindricalLensTesting.pdf)

**Reply by mjhoptics** [reply]
Hello Mike @mikejaris,
Thanks for the report and the info needed to reproduce the problem. I found the issue (the normal calculation was wrong for XToroids) and updated a fix to the develop branch. Hope to merge this back to main soon.
Sorry for the delay in getting to this.
Regards
Mike
